src/dfx/src/lib/canister_logs/log_visibility.rs

```diff
@@ -0,0 +1,144 @@
+                add_log_viewer: None,
+                remove_log_viewer: None,
+                set_log_viewer: None,
+            });
+        }
+
+        if let Some(log_viewer) = log_viewer {
+            return Some(LogVisibilityOpt {
+                log_visibility: None,
+                add_log_viewer: None,
+                remove_log_viewer: None,
+                set_log_viewer: Some(log_viewer.clone()),
+            });
+        }
+
+        None
+    }
+
+    pub fn to_log_visibility(
+        &self,
+        env: &dyn Environment,
+        current_status: Option<&StatusCallResult>,
+    ) -> DfxResult<LogVisibility> {
+        let logger = env.get_logger();
+
+        // For public and controllers.
+        if let Some(log_visibility) = self.log_visibility.as_ref() {
+            return Ok(log_visibility.clone());
+        }
+
+        // For setting viewers.
+        if let Some(viewers) = self.set_log_viewer.as_ref() {
+            let principals: DfxResult<Vec<_>> = viewers
+                .iter()
+                .map(|viewer| Ok(Principal::from_text(viewer).unwrap()))
+                .collect();
+
+            return Ok(LogVisibility::AllowedViewers(principals.unwrap()));
+        }
+
+        // Get the current viewer list for adding and removing, only for update-settings.
+        let mut current_visibility: Option<LogVisibility> = None;
+        let mut viewers = match current_status {
+            Some(status) => {
+                current_visibility = Some(status.settings.log_visibility.clone());
+                match &status.settings.log_visibility {
+                    LogVisibility::AllowedViewers(viewers) => viewers.clone(),
+                    _ => vec![],
+                }
+            }
+            None => vec![],
+        };
+
+        // Adding.
+        if let Some(added) = self.add_log_viewer.as_ref() {
+            if let Some(LogVisibility::Public) = current_visibility {
+                let msg = "Current log is public to everyone. Adding log reviewers will make the log only visible to the reviewers.";
+                ask_for_consent(msg)?;
+            }
+
+            for viewer in added {
+                let principal = Principal::from_text(viewer).unwrap();
+                if !viewers.iter().any(|x| *x == principal) {
+                    viewers.push(principal);
+                }
+            }
+        }
+
+        // Removing.
+        if let Some(removed) = self.remove_log_viewer.as_ref() {
+            if let Some(visibility) = &current_visibility {
+                match visibility {
+                    LogVisibility::Public | LogVisibility::Controllers => {
+                        return Err(anyhow!("Removing reviewers is not allowed with 'public' or 'controllers' log visibility."));
+                    }
+                    _ => (),
+                }
+            }
+            for viewer in removed {
+                let principal = Principal::from_text(viewer).unwrap();
```

Comment:
This is untrusted user input at this point, right? I don't think it's ok to just unwrap principals everywhere, but especially not if it hasn't been validated anywhere before. Can you try to find a better solution? It's probably not too hard when you use `add_log_viewer: Option<Vec<Principal>>` in the CLI opts

Reply:
Fixed at https://github.com/dfinity/sdk/pull/3921/commits/07b2ce497e7de94506012f7e669ff78d5779543d.